{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using pyensembl to get all gene info within the loci we are interested in\n",
    "# https://github.com/openvax/pyensembl\n",
    "from pyensembl import EnsemblRelease\n",
    "import pandas as pd\n",
    "## human reference genome\n",
    "data = EnsemblRelease(111)\n",
    "data.download()\n",
    "data.index()\n",
    "\n",
    "def get_gene_info(tad):\n",
    "    chr_num = tad.split('.')[0].split('r')[1]\n",
    "    start = int(tad.split('.')[1])\n",
    "    end = int(tad.split('.')[2])\n",
    "    genes = data.genes_at_locus(contig=chr_num, position=start, end=end)\n",
    "    gene_data = []\n",
    "    for gene in genes:\n",
    "        gene_info = {\n",
    "            \"Gene ID\": gene.gene_id,\n",
    "            \"Gene Name\": gene.gene_name,\n",
    "            \"Chromosome\": gene.contig,\n",
    "            \"Start\": gene.start,\n",
    "            \"End\": gene.end,\n",
    "            \"Strand\": gene.strand,\n",
    "            \"Biotype\": gene.biotype\n",
    "        }\n",
    "        gene_data.append(gene_info)\n",
    "\n",
    "    # Create a DataFrame\n",
    "    gene_df = pd.DataFrame(gene_data)\n",
    "    return gene_df\n",
    "\n",
    "def clean_cell_spe_gtex_pairs(gtex_pairs):\n",
    "    gtex_pairs['gene_name'] = gtex_pairs['gene_id'].str.split('.',expand=True)[0]\n",
    "    gtex_pairs = gtex_pairs[['variant_id','gene_name','tss_distance','slope','slope_se']]\n",
    "    return gtex_pairs\n",
    "\n",
    "def get_overlaped_pairs(chr_tad_gene,cell_spe_gtex_pairs):    \n",
    "    chr_overlap = chr_tad_gene.merge(cell_spe_gtex_pairs,\n",
    "                                       left_on= 'Gene ID',\n",
    "                                       right_on = 'gene_name',\n",
    "                                      how='inner')\n",
    "    return chr_overlap\n",
    "\n",
    "# get the index num of each egene and eqtl\n",
    "def get_index_num(df,tad):    \n",
    "    ## get the variant index number in the tad\n",
    "    df['variant_loc'] = df['variant_id'].str.split('_',expand=True)[1]\n",
    "    tad_start = int(tad.split('.')[1])\n",
    "    df['variant_index'] = ((df['variant_loc'].astype(int)-tad_start)/5000).astype(int)\n",
    "    ## get the gene TSS index number in the tad\n",
    "    df_minus_strand = df[df['Strand']==\"-\"]\n",
    "    df_minus_strand['TSS'] = df_minus_strand['End']\n",
    "    df_plus_strand = df[df['Strand']==\"+\"]\n",
    "    df_plus_strand['TSS'] = df_plus_strand['Start']\n",
    "    df_TSS = df_plus_strand.append(df_minus_strand)\n",
    "    df_TSS['TSS_index'] = ((df_TSS['TSS'].astype(int)-tad_start)/5000).astype(int)\n",
    "    df_TSS = df_TSS[['gene_name','variant_id','slope','variant_index','TSS_index']]\n",
    "    return df_TSS\n",
    "\n",
    "def get_lrange_pairs(cleaned_pairs):    \n",
    "    df = cleaned_pairs\n",
    "    df_res = df[(df['TSS_index'] != df['variant_index'])]\n",
    "    df_res = df_res[(df_res['variant_index'] >= 0)]\n",
    "    df_res = df_res.reset_index(drop=True)\n",
    "    return df_res\n",
    "\n",
    "def get_phy_pairs(bead_physical_contacted_df,chr_clean_pairs):\n",
    "    phy_pairs = bead_physical_contacted_df.merge(chr_clean_pairs,\n",
    "                                left_on = ['egenes','eQTLs'],\n",
    "                                right_on = ['TSS_index','variant_index'],\n",
    "                                how='inner')\n",
    "    phy_pairs = phy_pairs.drop(columns=['egenes', 'eQTLs']).reset_index(drop=True)\n",
    "    phy_pairs = phy_pairs[['gene_name','variant_id']]\n",
    "    return phy_pairs\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the gene info in chr2:230805000 - 231690000\n",
    "\n",
    "tad = 'chr2.230805000.231690000'\n",
    "chr_num = tad.split('.')[0].split('r')[1]\n",
    "tad_start = int(tad.split('.')[1])\n",
    "tad_end = int(tad.split('.')[2])\n",
    "\n",
    "chr2_tad_gene = get_gene_info(tad)\n",
    "\n",
    "# load all egene-eQTL pairs data downloaded from GTEx \n",
    "GM_gtex_pairs = pd.read_csv('../sample_data/GTEx_data/Cells_EBV-transformed_lymphocytes.v8.signif_variant_gene_pairs.txt',\n",
    "                        sep=\"\\t\")\n",
    "\n",
    "# clean the egene-eQTL pairs data to get the useful data\n",
    "GM_gtex_pairs_clean = clean_cell_spe_gtex_pairs(GM_gtex_pairs)\n",
    "\n",
    "\n",
    "# get the overlaped pairs in selected loci \n",
    "GM_chr2_tad_pairs = get_overlaped_pairs(chr2_tad_gene,GM_gtex_pairs_clean)\n",
    "\n",
    "GM_chr2_clean_pairs = get_index_num(GM_chr2_tad_pairs,tad)\n",
    "\n",
    "# get rid of eGene and eQTL located on the same beads. \n",
    "GM_chr2_lrange_clean_pairs = get_lrange_pairs(GM_chr2_clean_pairs)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "### get the physical contacted beads between eGene promoter and the other beads\n",
    "\n",
    "# load the distance dataframe for all chains\n",
    "GM_dist_df  = pd.read_csv('../sample_data/chain_dist_df/all_GM_dist_df.csv')\n",
    "\n",
    "dist_thrd = 100\n",
    "eGene_promoter_index = GM_chr2_lrange_clean_pairs.TSS_index.unique()\n",
    "\n",
    "promoter_list = []\n",
    "eqtl_list = []\n",
    "\n",
    "# check the median distance between the promoter of eGene and or other beads\n",
    "for p in eGene_promoter_index:\n",
    "    for eQTL in range(int((tad_end-tad_start)/5000)):\n",
    "        dist_data = GM_dist_df[(GM_dist_df.i==p) & (GM_dist_df.j==eQTL)].iloc[:,2:].T\n",
    "        dist_data_median = dist_data.median().item()\n",
    "\n",
    "        if dist_data_median < dist_thrd:\n",
    "            promoter_list.append(p)\n",
    "            eqtl_list.append(eQTL)\n",
    "GM_bead_physical_contacted_df = pd.DataFrame(list(zip(promoter_list,eqtl_list)),columns=['egenes','eQTLs'])\n",
    "\n",
    "### merge with GTEx data to get physical contacted eGene-eQTL pairs\n",
    "chr2_GM_phy_pairs = get_phy_pairs(GM_bead_physical_contacted_df,GM_chr2_lrange_clean_pairs)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
